src/mame/drivers/uknc.cpp

```diff
@@ -9,97 +9,946 @@ Graphics 640x288 pixels.
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(
+			YRGB(BIT(aux, i * 4 + 2), BIT(aux, i * 4 + 3), BIT(m_video.control, 18)),
+			YRGB(BIT(aux, i * 4 + 1), BIT(aux, i * 4 + 3), BIT(m_video.control, 17)),
+			YRGB(BIT(aux, i * 4    ), BIT(aux, i * 4 + 3), BIT(m_video.control, 16)))
+		);
+	}
+}
+
+void uknc_state::set_palette_ygrb(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(
+			YRGB(BIT(aux, i * 4 + 1), BIT(aux, i * 4 + 3), BIT(m_video.control, 17)),
+			YRGB(BIT(aux, i * 4 + 2), BIT(aux, i * 4 + 3), BIT(m_video.control, 18)),
+			YRGB(BIT(aux, i * 4    ), BIT(aux, i * 4 + 3), BIT(m_video.control, 16)))
+		);
+	}
+}
+
+void uknc_state::set_palette__rgb(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(
+			YRGB(BIT(aux, i * 4 + 2), !BIT(aux, i * 4 + 3), 1),
+			YRGB(BIT(aux, i * 4 + 1), !BIT(aux, i * 4 + 3), 1),
+			YRGB(BIT(aux, i * 4    ), !BIT(aux, i * 4 + 3), 1))
+		);
+	}
+}
+
+void uknc_state::set_palette_gray(uint32_t aux)
+{
+	for (int i = 0; i < 8; i++)
+	{
+		m_palette->set_pen_color(i, rgb_t(i * 255 / 8, i * 255 / 8, i * 255 / 8));
+	}
+}
+
+#undef YGRB
+
+
+void uknc_state::update_displaylist()
+{
+	int row, wide = 0;
+	uint16_t addr = 0270, nextaddr;
+	bool cursor = false, cursor_type = false;
+	int cursor_color = 0, cursor_octet = 0, cursor_pixel = 0;
+	scanline_t *scanline;
+
+	for (row = 0; row < UKNC_DISP_VERT; row++)
+	{
+		scanline = &m_scanlines[row];
+		scanline->addr = m_subcpu->space(AS_PROGRAM).read_word(addr);
+		nextaddr = m_subcpu->space(AS_PROGRAM).read_word(addr + 2);
+
+		if (BIT(nextaddr, 0))
+			cursor = !cursor;
+
+		scanline->control = scanline->color = false;
+		scanline->cursor = cursor;
+
+		if (wide)
+		{
+			scanline->aux = m_subcpu->space(AS_PROGRAM).read_word(addr - 4) |
+				(m_subcpu->space(AS_PROGRAM).read_word(addr - 2) << 16);
+			if (wide == 1)
+			{
+				scanline->control = true;
+				cursor_type = BIT(scanline->aux, 4);
+				cursor_color = scanline->aux & 15;
+				cursor_pixel = (scanline->aux >> 4) & 3;
+				cursor_octet = (scanline->aux >> 8) & 127;
+			}
+			else
+			{
+				scanline->color = true;
+			}
+		}
+
+		scanline->cursor_type = cursor_type;
+		scanline->cursor_color = cursor_color;
+		scanline->cursor_pixel = cursor_pixel;
+		scanline->cursor_octet = cursor_octet;
+
+		if (BIT(nextaddr, 1))
+		{
+			wide = BIT(nextaddr, 2) + 1;
+			nextaddr &= ~7;
+			nextaddr += 4;
+		}
+		else
+		{
+			wide = 0;
+			nextaddr &= ~3;
+		}
+
+		addr = nextaddr;
+	}
+}
+
+/*
+ * aux bits, word 0:
+ *
+ * 0-3	cursor color, YRGB
+ * 4	cursor type, 0: character, 1: graphics
+ * 5-7	graphics cursor position in octet
+ * 8-14	cursor position in line
+ * 15	-
+ *
+ * aux bits, word 1:
+ *
+ * 16-18	RGB intensity
+ * 19	-
+ * 20-21	scaling
+ */
+void uknc_state::draw_scanline(uint32_t *p, scanline_t *scanline)
+{
+	int i, start, wide = 0, pixel_x;
+	const pen_t *pen = m_palette->pens();
```

Comment:
Same for these – they should be scoped down.

Reply:
changed.